app/adapters/bagit/storage_adapter.rb

```diff
@@ -0,0 +1,81 @@
+# frozen_string_literal: true
+module Bagit
+  class StorageAdapter
+    attr_reader :base_path
+    def initialize(base_path:)
+      @base_path = Pathname.new(base_path)
+    end
+
+    def for(bag_id:)
+      Instance.new(base_path: base_path, bag_id: bag_id.to_s)
+    end
+
+    class Instance
+      attr_reader :base_path, :bag_id
+      def initialize(base_path:, bag_id:)
+        @base_path = base_path
+        @bag_id = bag_id
+      end
+
+      def upload(file:, original_filename:, resource: nil)
+        FileUtils.mkdir_p(data_path)
+        new_path = data_path.join("#{generate_id}-#{original_filename}")
+        old_path = file.try(:disk_path) || file.path
+        FileUtils.cp(old_path, new_path)
+        output_file = find_by(id: Valkyrie::ID.new("bag://#{new_path.relative_path_from(base_path)}"))
+        create_manifests(output_file)
+        output_file
+      end
+
+      def create_manifests(file)
+        checksums = file.checksum(digests: [Digest::SHA1.new, Digest::MD5.new, Digest::SHA256.new])
```

Comment:
We talked about this on Slack, but I'm putting a note here so I don't forget: it would be better to use the existing checksums from the file metadata — both because it's faster and because it would catch any corruption that might happen when copying the files.